lib/ui/painting/image_descriptor.cc

```diff
@@ -0,0 +1,112 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/image_descriptor.h"
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/trace_event.h"
+#include "flutter/lib/ui/painting/codec.h"
+#include "flutter/lib/ui/painting/image_decoder.h"
+#include "flutter/lib/ui/painting/multi_frame_codec.h"
+#include "flutter/lib/ui/painting/single_frame_codec.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/tonic/dart_binding_macros.h"
+#include "third_party/tonic/logging/dart_invoke.h"
+
+namespace flutter {
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, ImageDescriptor);
+
+#define FOR_EACH_BINDING(V)            \
+  V(ImageDescriptor, initRaw)          \
+  V(ImageDescriptor, instantiateCodec) \
+  V(ImageDescriptor, width)            \
+  V(ImageDescriptor, height)           \
+  V(ImageDescriptor, bytesPerPixel)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+void ImageDescriptor::RegisterNatives(tonic::DartLibraryNatives* natives) {
+  natives->Register(
+      {{"ImageDescriptor_initEncoded", ImageDescriptor::initEncoded, 3, true},
+       FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
+}
+
+void ImageDescriptor::initEncoded(Dart_NativeArguments args) {
+  Dart_Handle callback_handle = Dart_GetNativeArgument(args, 2);
+  if (!Dart_IsClosure(callback_handle)) {
```

Comment:
We seem to have gotten into a situation again where we are using both callbacks and return codes for results. I suggest using one of the other.

How about this: Make `initEncoded` a static method that takes a callback that returns a descriptor or an error message (or null for the generic "cannot decode" message). You can return early if the callback itself is unspecified.

Reply:
This is a `dart:ui` thing. We could refactor this, but the Dart signature of this method (and a couple others) expects a nullable string to be returned if there's an error. 

The basic problem here is that we want to avoid doing a Dart throw from the C++ side, since that causes other problems.

I'm not sure how to structure the callback you're suggesting on the Dart side any better than what we have today - at the end of ita ll, we'd want a String on error and an ImageDescriptor on success.